# Лабораторная работа №1
## "Предварительный анализ данных"

### Подготовка данных

В данной лабораторной работе необходимо загрузить и обработать датасет, а также выполнить задания по варианту с помощью библиотеки `pandas` в `Python`. Для начала была импортирована данная библиотека с помощью функции `import`, затем был загружен датасет `clients.csv`

In [373]:
import pandas as pd

df = pd.read_csv('clients.csv',sep=';')

Были выведены первые 20 строк с помощью функции `df.head(20)`

In [374]:
df.head(20)

ID  Year_Birth   Education Marital_Status   Income  Kidhome Dt_Customer  \
0   5524        1957  Graduation         Single  58138.0      0.0  04.09.2012   
1   2174        1954  Graduation         Single  46344.0      1.0  08.03.2014   
2   4141        1965  Graduation       Together  71613.0      0.0  21.08.2013   
3   6182        1984  Graduation       Together  26646.0      1.0  10.02.2014   
4   5324        1981         PhD        Married  58293.0      1.0  19.01.2014   
5   7446        1967      Master       Together  62513.0      0.0  09.09.2013   
6    965        1971  Graduation       Divorced  55635.0      0.0  13.11.2012   
7   6177        1985         PhD        Married  33454.0      1.0  08.05.2013   
8   4855        1974         PhD       Together  30351.0      1.0  06.06.2013   
9   5899        1950         PhD       Together   5648.0      1.0  13.03.2014   
10  1994        1983  Graduation        MARRIED      NaN      NaN         NaN   
11   387        1976       Basic        Married   7500.0      0.0  13.11.2012   
12  2125        1959  Graduation       Divorced  63033.0      0.0  15.11.2013   
13  8180        1952      Master       Divorced  59354.0      1.0  15.11.2013   
14  2569        1987  Graduation        Married  17323.0      0.0  10.10.2012   
15  2114        1946         PhD          SINGL  82800.0      0.0  24.11.2012   
16  9736        1980  Graduation        Married  41850.0      1.0  24.12.2012   
17  4939        1946  Graduation       Together  37760.0      0.0  31.08.2012   
18  6565        1949      Master        Married  76995.0      0.0  28.03.2013   
19  9360        1982  Graduation        Married  37040.0      0.0  08.08.2012   

    NumDealsPurchases  
0                 3.0  
1                 2.0  
2                 1.0  
3                 2.0  
4                 5.0  
5                 2.0  
6                 4.0  
7                 2.0  
8                 1.0  
9                 1.0  
10                NaN  
11                1.0  
12                1.0  
13                3.0  
14                1.0  
15                1.0  
16                3.0  
17                2.0  
18                2.0  
19                1.0

Предметной областью по варианту является журнал клиентов с такой информацией о них:
+ уникальный идентификатор
+ год рождения
+ уровень образования (базовый, выпускник, магистр, доктор наук)
+ семейное положение (холостой, в отношениях, женатый, разведён, вдовый)
+ семейная прибыль
+ количество детей
+ дата регистрации в компании
+ количество покупок

Все эти характеристики отображены в столбцах данного датасета. Чтобы показать все столбцы, используется функция `columns`

In [375]:
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Dt_Customer', 'NumDealsPurchases'],
      dtype='object')

Когда все столбцы были выведены, есть необходимость заменить названия столбцов, так как некоторые не отображают контекст (к примеру, `Kidhome`) или имеют в названии ошибку (в `NumDealsPurchases` слова deals и purchases имеют схожий смысл, а также написание не похоже на остальные столбцы)

In [376]:
df = df.rename(columns= {'Kidhome':'Children'})
df = df.rename(columns= {'NumDealsPurchases':'Purchases'})
df = df.rename(columns= {'Year_Birth': 'Birth_Year'})
df = df.rename(columns= {'Dt_Customer': 'Reg_Date'})

print(df.columns)

Index(['ID', 'Birth_Year', 'Education', 'Marital_Status', 'Income', 'Children',
       'Reg_Date', 'Purchases'],
      dtype='object')


Для отображения информации о датасете используется функция `df.info()`

In [377]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 796 entries, 0 to 795
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID              796 non-null    int64  
 1   Birth_Year      796 non-null    int64  
 2   Education       796 non-null    object 
 3   Marital_Status  796 non-null    object 
 4   Income          784 non-null    float64
 5   Children        795 non-null    float64
 6   Reg_Date        795 non-null    object 
 7   Purchases       795 non-null    float64
dtypes: float64(3), int64(2), object(3)
memory usage: 49.9+ KB


Всего записей в этом датасете 796, однако в столбцах Income, Children, Reg_Date и Purchases наблюдаются меньше записей, так как там есть значения `NaN`. Чтобы вывести все записи, имеющие хотя бы одно нулевое значение, используется функция `.isnull()`. Функция `.any(axis=1)` указывает на поиск ненулевых значений именно по строкам.

In [378]:
df_nan = df[df.isnull().any(axis=1)]
display(df_nan)

ID  Birth_Year   Education Marital_Status  Income  Children  \
10   1994        1983  Graduation        MARRIED     NaN       NaN   
26   5255        1986  Graduation         Single     NaN       1.0   
41   7281        1959         PhD         Single     NaN       0.0   
45   7244        1951  Graduation         Single     NaN       2.0   
55   8557        1982  Graduation         Single     NaN       1.0   
83   8996        1957         PhD        Married     NaN       2.0   
84   9235        1957  Graduation         Single     NaN       1.0   
85   5798        1973      Master       Together     NaN       0.0   
116  8268        1961         PhD        Married     NaN       0.0   
121  1295        1963  Graduation        Married     NaN       0.0   
286  2437        1989  Graduation        Married     NaN       0.0   
293  2863        1970  Graduation         Single     NaN       1.0   

       Reg_Date  Purchases  
10          NaN        NaN  
26   20.02.2013        0.0  
41   05.11.2013        1.0  
45   01.01.2014        3.0  
55   17.06.2013        2.0  
83   19.11.2012       12.0  
84   27.05.2014        1.0  
85   23.11.2013        1.0  
116  11.07.2013        3.0  
121  11.08.2013        1.0  
286  03.06.2013        1.0  
293  23.08.2013        6.0

Было принято решение удалить все записи, содержащие в себе хотя бы одно значение `NaN`, так как в дальнейшем отсутствие данных повлияло бы на значение среднего дохода семьи. За удаление строк с нулевыми значениями отвечает функция `.dropna()`. После этого была проведена проверка нулевых значений в столбцах с помощью функции `.isna()`, `.sum()` показывает сумму таких значений. Везде написано 0

In [379]:
df = df.dropna()
df.isna().sum()

ID                0
Birth_Year        0
Education         0
Marital_Status    0
Income            0
Children          0
Reg_Date          0
Purchases         0
dtype: int64

Также вывод функции `.info()` показал типы данных в столбцах. Столбцы `Children` и `Purchases` имеют тип `float`, однако значения в данных столбцах могут быть только дискретными, поэтому необходимо задать данным столбцам тип данных `int` с помощью функции `.astype()`

In [380]:
df['Children'] = df['Children'].astype(int)
df['Purchases'] = df['Purchases'].astype(int)

Также столбец `Reg_Date` имеет тип данных `object`, но так как это столбец с датой, то данному столбцу необходимо присвоить тип данных `datetime` для правильного отображения с помощью функции `pd.to_datetime()` и задать формат дд.мм.гг.

In [381]:
df['Reg_Date'] = pd.to_datetime(df['Reg_Date'], format='%d.%m.%Y')

 Теперь можно вывести информацию о датасете с изменёнными типами данных.

In [382]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 784 entries, 0 to 795
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   ID              784 non-null    int64         
 1   Birth_Year      784 non-null    int64         
 2   Education       784 non-null    object        
 3   Marital_Status  784 non-null    object        
 4   Income          784 non-null    float64       
 5   Children        784 non-null    int64         
 6   Reg_Date        784 non-null    datetime64[ns]
 7   Purchases       784 non-null    int64         
dtypes: datetime64[ns](1), float64(1), int64(4), object(2)
memory usage: 55.1+ KB


Метод `.describe()` используется для получения статистического обзора данных. Он предоставляет такие статистические показатели, как среднее значение, стандартное отклонение, минимальное и максимальное значения, а также квартильные значения (25%, 50%, и 75%).
Столбец `Reg_Date` тоже содержит числовые значения, но в виде даты, поэтому он был исключён из вывода, так как использовать его было бы некорректно.

In [383]:
description = df.drop(columns='Reg_Date').describe()
display(description)

ID   Birth_Year        Income    Children   Purchases
count    784.000000   784.000000     784.00000  784.000000  784.000000
mean    5628.014031  1968.343112   53130.07398    0.434949    2.307398
std     3279.536751    12.013759   21818.56876    0.542783    1.914827
min        0.000000  1899.000000    2447.00000    0.000000    0.000000
25%     2853.000000  1959.000000   36141.75000    0.000000    1.000000
50%     5558.500000  1969.500000   52372.50000    0.000000    2.000000
75%     8594.250000  1977.000000   69293.25000    1.000000    3.000000
max    11191.000000  1995.000000  162397.00000    2.000000   15.000000

Далее необходимо проверить датасет на явные и неявные дубликаты. Для начала были проверены столбцы `Marital_Status` и `Education` на наличие записей с одинаковыми значениями, но разные по написанию (регистр или синоним) с помощью функции `.unique()`

In [384]:
print(df['Marital_Status'].unique())
print(df['Education'].unique())

['Single' 'Together' 'Married' 'Divorced' 'SINGL' 'MARRIED' 'Widow'
 'Alone']
['Graduation' 'PhD' 'Master' 'Basic']


В столбце `Education` нет каких-либо дублирующихся значений, однако в столбце `Marital_Status` есть значения `SINGL` и `Alone`, которые дублируют `Single`, и `MARRIED`, дублирующий `Married`. Приведение записей с такими значениями в нормальный вид можно сделать с помощью функции `.replace()`

In [385]:
dubles = ['SINGL','Alone']
df['Marital_Status'] = df['Marital_Status'].replace(dubles,'Single')
df['Marital_Status'] = df['Marital_Status'].replace('MARRIED','Married')
print(df['Marital_Status'].unique())

['Single' 'Together' 'Married' 'Divorced' 'Widow']


С помощью функции `.duplicated()` можно проверить, какие записи полностью повторяют уже существующую запись, но таким образом он выведет весь датасет со значениями `True/False`, поэтому для упрощения был добавлен метод `.sum()`

In [386]:
print(df.duplicated().sum())

4


Всего получилось 4 дубликатов, которые необходимо удалить с помощью функции `.drop_duplicates()`

In [387]:
df = df.drop_duplicates()

В результате датасет был обработан и приведён в подходящий для анализа вид. Столбцы были переименованы и у некоторых из них были изменены типы данных. Датасет был очищен от дубликатов и от записей с неизвестными значениями.

### Задание 1
_Группировка - тип образования по каждому семейному статусу (marital_status)._

Группировка данных происходит с помощью метода `.groupby()`. Необходимо было сгруппировать данные по столбцам `Education` и `Marital_Status`, и вывести количество записей, соответствующих определённому семейному положению на всех уровнях образования, выведя только столбец `ID` с методом `.count()`

In [388]:
un = df.groupby(['Education','Marital_Status'])
group = un['ID'].count()
display(group)

Education   Marital_Status
Basic       Divorced            1
            Married             9
            Single              1
            Together            4
Graduation  Divorced           49
            Married           168
            Single             98
            Together           96
            Widow              21
Master      Divorced           12
            Married            57
            Single             32
            Together           41
            Widow               4
PhD         Divorced           23
            Married            68
            Single             41
            Together           51
            Widow               4
Name: ID, dtype: int64

Получилась таблица, сгруппированная по уровню образования, которая показывает количество клиентов с разным семейным положением на все уровнях образования

### Задание 2
_Группировка - тип образования по каждому семейному статусу (marital_status). Создать датафрейм. Переименовать столбец с количеством в “сount”. Отсортировать по убыванию столбца “count”._


Так как группировка точно такая же, то можно работать с таблицей из прошлого задания. Но теперь необходимо из этой таблицы создать датафрейм с помощью метода `pd.DataFrame()`. Самой таблице была присвоена переменная `group`, на основе которой был создан датафрейм. После этого столбец `ID` был переименован в столбец `Count` 

In [389]:
gr_df = pd.DataFrame(group)
gr_df = gr_df.rename(columns={'ID':'Count'}) 

Далее значения столбца `Count` были отсортированы по убыванию, используя метод `sort_values`. Стоит обратить внимание на то, что внутри метода пишется всегда атрибут `ascending`, даже если надо отсортировать по убыванию. Атрибут `ascending` принимает значения `True/False`

In [390]:
gr_df.sort_values(by='Count', ascending= False)

Count
Education  Marital_Status       
Graduation Married           168
           Single             98
           Together           96
PhD        Married            68
Master     Married            57
PhD        Together           51
Graduation Divorced           49
PhD        Single             41
Master     Together           41
           Single             32
PhD        Divorced           23
Graduation Widow              21
Master     Divorced           12
Basic      Married             9
           Together            4
Master     Widow               4
PhD        Widow               4
Basic      Single              1
           Divorced            1

В результате получилась отсортирована таблица по количеству записей с определённым семейным положением на различных уровнях образования

### Задание 3
_Сводная таблица (pivot_table) - средний доход семьи по семейному положению. Отсортировать по убыванию. Округлить до двух знаков._


Метод `.pivot_table()` используется для создания сводных таблиц, которые позволяют агрегировать данные и представлять их в удобном формате. Основные атрибуты метода:
+ `values`: Столбец или столбцы для агрегирования
+ `index`: Столбец или столбцы, которые будут использоваться в качестве индексов строк
+ `columns`: Столбец или столбцы, которые будут использоваться в качестве заголовков столбцов
+ `aggfunc`: Функция или список функций для агрегации

В данном задании необходимо создать сводную таблицу среднего дохода только по семейному положению, поэтому из списка используются только атрибуты `values`, `index` и `aggfunc`. За значения берётся доход, за индексы - семейное положение и функция усреднения.
Также значения были отсортированы по убыванию и округлены до 2 знаков с помощью метода `.round(2)`

In [391]:
pivot = df.pivot_table(values='Income', index='Marital_Status', aggfunc='mean')
display(pivot.sort_values(by='Income',ascending= False).round(2))

Income
Marital_Status          
Widow           55452.45
Divorced        54568.16
Together        54502.84
Married         52898.34
Single          50934.03

Получившаяся таблица показывает среднее значение дохода семьи по семейному положению.

### Задание 4
_Сводная таблица (pivot_table) - средний доход семьи по семейному положению - строки и уровню образованию - столбцы. Отсортировать по возрастанию marital_status. Округлить до двух знаков._


В данном задании необходимо сделать такую же сводную таблицу, но с использованием атрибута `columns`, в качестве которого используется столбец `Education`. Значения были отсортированы по возрастанию индексов `Marital_Status` и округлены до 2 знаков.

In [392]:
pivot_sq = df.pivot_table(values='Income', index='Marital_Status',columns='Education', aggfunc='mean')
display(pivot_sq.sort_values(by= 'Marital_Status', ascending= True).round(2))

Education          Basic  Graduation    Master       PhD
Marital_Status                                          
Divorced         9548.00    57420.98  49405.83  53141.22
Married         22018.22    51078.81  55391.37  59391.00
Single          21063.00    50325.66  50831.16  53197.05
Together        22437.50    52922.06  51009.98  62801.33
Widow                NaN    55146.81  47164.00  65345.50

Данная таблица отображает средние значения дохода семьи по разным уровням образования и семейному положению клиентов. Также в данной таблице можно заметить значение NaN. Это обусловлено тем, что в датасете не было заявлено записей вдовцов с базовым образованием.

### Вывод

В данной лабораторной работе была проделана работа по обработке датасета, исключив ненужные записи и заменив некоторые значения, а также анализу данных с помощью группировки и создания сводной таблицы. Были освоены основные методы библиотеки `pandas` и варианты их использования:
1. Для вывода информации о датасете были использованы методы `.info()`, `.describe()`
2. Для вывода информации об определённых аспектах датасета были использованы методы `df.column`, `.isna()`, `.isnull()`, `.unique()`, `.duplicated()`
3. Для замены данных в датасете были использованы методы `.rename()` - для столбцов или значений внутри столбцов, `.astype()` и `pd.to_datetime` - для изменения типа данных какого-либо столбца

Также были использованы методы `.groupby()` для группировки значений, `sort_values()` для их сортировки, `.pivot_table()` для создания сводной таблицы и `pd.DataFrame()` для создания датафрейма из таблицы.


### Дополнительное задание
_В столбце "Income" выделить 4 категории, сгруппировать по количеству записей. Выделить, какая категория имеет больше всего записей, фильтровать по этой категории и вывести отсортированную по убыванию сводную таблицу со столбцами "Marital_Status", "Education", средние значения "Children"_

In [393]:
bounds = [0,25000,50000,75000,100000]
categories = ['Low', 'Medium', 'High', 'Very High']

df['IncomeGroup'] = pd.cut(df['Income'], bins= bounds, labels= categories)

df_cat = df.groupby('IncomeGroup', observed= True)
df_catstat = df_cat['ID'].count()
display(df_catstat)

IncomeGroup
Low           77
Medium       279
High         289
Very High    127
Name: ID, dtype: int64

In [394]:
df_filter = df.query('IncomeGroup == "High"')
display(df_filter)

ID  Birth_Year   Education Marital_Status   Income  Children  \
0     5524        1957  Graduation         Single  58138.0         0   
2     4141        1965  Graduation       Together  71613.0         0   
4     5324        1981         PhD        Married  58293.0         1   
5     7446        1967      Master       Together  62513.0         0   
6      965        1971  Graduation       Divorced  55635.0         0   
..     ...         ...         ...            ...      ...       ...   
783   3643        1974      Master        Married  63206.0         0   
784    851        1972      Master        Married  57136.0         1   
786  10128        1958  Graduation       Together  53977.0         0   
789   9451        1965  Graduation        Married  73538.0         0   
791   2853        1980  Graduation         Single  51766.0         1   

      Reg_Date  Purchases IncomeGroup  
0   2012-09-04          3        High  
2   2013-08-21          1        High  
4   2014-01-19          5        High  
5   2013-09-09          2        High  
6   2012-11-13          4        High  
..         ...        ...         ...  
783 2014-03-20          1        High  
784 2013-07-09          9        High  
786 2013-06-08          5        High  
789 2012-11-25          3        High  
791 2014-03-11          2        High  

[289 rows x 9 columns]

In [395]:
df_high = pd.DataFrame(df_filter)
pivot_filter = df_high.pivot_table(values= 'Children', index= ['Marital_Status', 'Education'], aggfunc= 'mean')
display(pivot_filter)

Children
Marital_Status Education           
Divorced       Graduation  0.052632
               Master      0.400000
               PhD         0.200000
Married        Graduation  0.262295
               Master      0.227273
               PhD         0.233333
Single         Graduation  0.133333
               Master      0.222222
               PhD         0.142857
Together       Graduation  0.125000
               Master      0.294118
               PhD         0.111111
Widow          Graduation  0.111111
               Master      0.000000
               PhD         0.000000